cnf-tests/testsuites/pkg/machineconfigpool/machineconfigpool.go

```diff
@@ -135,3 +144,128 @@ func DecodeMCYaml(mcyaml string) (*mcov1.MachineConfig, error) {
+		},
+
+		Spec: mcov1.MachineConfigPoolSpec{
+			MachineConfigSelector: &metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      "machineconfiguration.openshift.io/role",
+					Operator: metav1.LabelSelectorOpIn,
+					Values:   []string{name, "worker"},
+				}},
+			},
+			Paused: false,
+			NodeSelector: &metav1.LabelSelector{
+				MatchLabels: newNodeRoleSelector,
+			},
+		},
+	}
+
+	kubeletConfig := &mcov1.KubeletConfig{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+		Spec:       *spec.DeepCopy(),
+	}
+
+	// Link the KubeletConfig to the MCP
+	kubeletConfig.Spec.MachineConfigPoolSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"machineconfiguration.openshift.io/role": name},
+	}
+
+	// Create the KubeletConfig
+	_, err := controllerutil.CreateOrUpdate(context.Background(), client.Client, kubeletConfig, func() error { return nil })
+	if err != nil {
+		return nilFn, err
+	}
+	klog.Infof("Created KubeletConfig %s", kubeletConfig.Name)
+
+	// Create MCP
+	_, err = controllerutil.CreateOrUpdate(context.Background(), client.Client, &mcp, func() error { return nil })
+	if err != nil {
+		return nilFn, err
+	}
+	klog.Infof("Created MachineConfigPool %s", mcp.Name)
+
+	// Following wait ensure the node is rebooted only once, as if we apply the MCP to
+	// the node before the KubeletConfig has been rendered, the node will reboot twice.
+	klog.Infof("Waiting for KubeletConfig to be rendered to MCP")
+	err = waitUntilKubeleConfigHasUpdatedTheMCP(name)
+	if err != nil {
+		return nilFn, err
+	}
+
+	// Move the node role to the new one
+	previousNodeRole := nodes.FindRoleLabel(node)
+	if previousNodeRole != "" {
+		err = nodes.RemoveRoleFrom(node.Name, previousNodeRole)
+		if err != nil {
+			return nilFn, err
+		}
+		klog.Infof("Removed role[%s] from node %s", previousNodeRole, node.Name)
+	}
+
+	err = nodes.AddRoleTo(node.Name, newNodeRole)
+	if err != nil {
+		return func() {
+			nodes.AddRoleTo(node.Name, previousNodeRole)
+			klog.Infof("Restored role[%s] on node %s", previousNodeRole, node.Name)
+		}, err
+	}
+	klog.Infof("Added role[%s] to node %s", newNodeRole, node.Name)
+
+	err = WaitForMCPStable(mcp)
+	if err != nil {
+		return func() {
+			nodes.AddRoleTo(node.Name, previousNodeRole)
+			nodes.RemoveRoleFrom(node.Name, newNodeRole)
+		}, err
+	}
+
+	return func() {
+		nodes.AddRoleTo(node.Name, previousNodeRole)
+		nodes.RemoveRoleFrom(node.Name, newNodeRole)
+		klog.Infof("Moved back node role from [%s] to [%s] on %s", newNodeRole, previousNodeRole, node.Name)
+
+		WaitForMCPStable(mcp)
+	}, nil
+}
+
+func waitUntilKubeleConfigHasUpdatedTheMCP(name string) error {
+	return wait.Poll(10*time.Second, 3*time.Minute, func() (bool, error) {
+
+		mcp, err := client.Client.MachineConfigPools().Get(context.Background(), name, metav1.GetOptions{})
+		if err != nil {
+			klog.Warningf("Error while waiting for MachineConfigPool[%s] to be updated: %v", name, err)
+			return false, nil
```

Comment:
There can be an API server error or some other transient error. Here we return nil because the `wait.Poll(...)` function stops looping if the callback returns an error. It's different than the gomega `Eventually` loop.